File: FFRv2/app/src/main/java/com/devingotaswitch/rankings/domain/RosterSettings.java
package com.devingotaswitch.rankings.domain;

import android.util.Log;

import com.devingotaswitch.utils.Constants;

import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

public class RosterSettings {

    private String id;
    private int qbCount;
    private int rbCount;
    private int wrCount;
    private int teCount;
    private int dstCount;
    private int kCount;
    private int benchCount;
    private Flex flex;

    private Set<String> validPositions;

    public RosterSettings() {
        this(UUID.randomUUID().toString(), Constants.ONE_STARTER, Constants.TWO_STARTERS, Constants.TWO_STARTERS, Constants.ONE_STARTER, Constants.ONE_STARTER,
                Constants.ONE_STARTER, Constants.BENCH_DEFAULT, new Flex());
    }

    public RosterSettings(int qbCt, int rbCt, int wrCt, int teCt, int dCt, int kCt,
                          int benchCt) {
        this(UUID.randomUUID().toString(), qbCt, rbCt, wrCt, teCt, dCt, kCt, benchCt, new Flex());
    }

    public RosterSettings(String id, int qbCt, int rbCt, int wrCt, int teCt, int dCt, int kCt,
                          int benchCt, Flex flex) {
        this.setId(id);
        this.setQbCount(qbCt);
        this.setRbCount(rbCt);
        this.setWrCount(wrCt);
        this.setTeCount(teCt);
        this.setDstCount(dCt);
        this.setkCount(kCt);
        this.setBenchCount(benchCt);
        this.setFlex(flex);

        this.validPositions = new HashSet<>();
        if (qbCt > 0 || flex.getQbrbwrteCount() > 0) {
            validPositions.add(Constants.QB);
        }
        if (rbCt > 0 || flex.getRbwrCount() > 0 || flex.getRbwrteCount() > 0 || flex.getRbteCount() > 0 || flex.getQbrbwrteCount() > 0) {
            validPositions.add(Constants.RB);
        }
        if (wrCt > 0 || flex.getRbwrCount() > 0 || flex.getRbwrteCount() > 0 || flex.getWrteCount() > 0 || flex.getQbrbwrteCount() > 0) {
            validPositions.add(Constants.WR);
        }
        if (teCt > 0 || flex.getRbwrteCount() > 0 || flex.getRbteCount() > 0 || flex.getWrteCount() > 0 || flex.getQbrbwrteCount() > 0) {
            validPositions.add(Constants.TE);
        }
        if (dCt > 0) {
            validPositions.add(Constants.DST);
        }
        if (kCt > 0) {
            validPositions.add(Constants.K);
        }
    }

    public String getId() {
        return id;
    }

    private void setId(String id) {
        this.id = id;
    }

    public int getQbCount() {
        return qbCount;
    }

    public void setQbCount(int qbCount) {
        this.qbCount = qbCount;
    }

    public int getRbCount() {
        return rbCount;
    }

    public void setRbCount(int rbCount) {
        this.rbCount = rbCount;
    }

    public int getWrCount() {
        return wrCount;
    }

    public void setWrCount(int wrCount) {
        this.wrCount = wrCount;
    }

    public int getTeCount() {
        return teCount;
    }

    public void setTeCount(int teCount) {
        this.teCount = teCount;
    }

    public int getDstCount() {
        return dstCount;
    }

    public void setDstCount(int dstCount) {
        this.dstCount = dstCount;
    }

    public int getkCount() {
        return kCount;
    }

    public void setkCount(int kCount) {
        this.kCount = kCount;
    }

    public int getBenchCount() {
        return benchCount;
    }

    public void setBenchCount(int benchCount) {
        this.benchCount = benchCount;
    }

    public Flex getFlex() {
        return flex;
    }

    public void setFlex(Flex flex) {
        this.flex = flex;
    }

    public int getRosterSize() {
        int size = qbCount + rbCount + wrCount + teCount + dstCount + kCount + benchCount;
        if (flex != null) {
            size += flex.getQbrbwrteCount() + flex.getRbwrteCount() + flex.getRbteCount() + flex.getRbwrCount() + flex.getWrteCount();
        }
        return size;
    }

    public int getNumberStartedOfPos(String position) {
        int total = 0;
        if (Constants.QB.equals(position)) {
            total = getQbCount();
            if (flex != null) {
                // Assume all qbs
                total += flex.getQbrbwrteCount();
            }
        } else if (Constants.RB.equals(position)) {
            total = getRbCount();
            if (flex != null) {
                total += flex.getRbteCount();
                total += flex.getRbwrCount();
                total += flex.getRbwrteCount();
            }
        } else if (Constants.WR.equals(position)) {
            total = getWrCount();
            if (flex != null) {
                total += flex.getWrteCount();
                total += flex.getRbwrCount();
                total += flex.getRbwrteCount();
            }
        } else if (Constants.TE.equals(position)) {
            total = getTeCount();
            // Assume no flex spot would go to a te
        } else if (Constants.DST.equals(position)) {
            total = getDstCount();
        } else if (Constants.K.equals(position)) {
            total = getkCount();
        }

        return total;
    }

    public boolean isPositionValid(String pos) {
        return validPositions.contains(pos);
    }

    public static class Flex {

        private int rbwrCount;
        private int rbwrteCount;
        private int rbteCount;
        private int wrteCount;
        private int qbrbwrteCount;

        public Flex() {
            this(Constants.ONE_STARTER, Constants.NO_STARTERS, Constants.NO_STARTERS, Constants.NO_STARTERS, Constants.NO_STARTERS);
        }

        public Flex(int rbwr, int rbwrte, int rbte, int wrte, int qbrbwrte) {
            this.setRbwrCount(rbwr);
            this.setRbteCount(rbte);
            this.setRbwrteCount(rbwrte);
            this.setWrteCount(wrte);
            this.setQbrbwrteCount(qbrbwrte);
        }

        public int getRbwrCount() {
            return rbwrCount;
        }

        public void setRbwrCount(int rbwrCount) {
            this.rbwrCount = rbwrCount;
        }

        public int getRbwrteCount() {
            return rbwrteCount;
        }

        public void setRbwrteCount(int rbwrteCount) {
            this.rbwrteCount = rbwrteCount;
        }

        public int getRbteCount() {
            return rbteCount;
        }

        public void setRbteCount(int rbteCount) {
            this.rbteCount = rbteCount;
        }

        public int getWrteCount() {
            return wrteCount;
        }

        public void setWrteCount(int wrteCount) {
            this.wrteCount = wrteCount;
        }

        public int getQbrbwrteCount() {
            return qbrbwrteCount;
        }

        public void setQbrbwrteCount(int qbrbwrteCount) {
            this.qbrbwrteCount = qbrbwrteCount;
        }
    }
}

File: FFRv2/app/src/main/java/com/devingotaswitch/rankings/domain/RosterSettings.kt
package com.devingotaswitch.rankings.domain

import com.devingotaswitch.utils.Constants
import java.util.*

class RosterSettings @JvmOverloads constructor(id: String = UUID.randomUUID().toString(),
                                               qbCt: Int = Constants.ONE_STARTER, rbCt: Int = Constants.TWO_STARTERS,
                                               wrCt: Int = Constants.TWO_STARTERS, teCt: Int = Constants.ONE_STARTER,
                                               dCt: Int = Constants.ONE_STARTER, kCt: Int = Constants.ONE_STARTER,
                                               benchCt: Int = Constants.BENCH_DEFAULT, flex: Flex = Flex()) {
    var id: String = ""
        private set
    var qbCount = 0
    var rbCount = 0
    var wrCount = 0
    var teCount = 0
    var dstCount = 0
    var kCount = 0
    var benchCount = 0
    var flex: Flex? = null
    private val validPositions: MutableSet<String>

    constructor(qbCt: Int, rbCt: Int, wrCt: Int, teCt: Int, dCt: Int, kCt: Int,
                benchCt: Int) : this(UUID.randomUUID().toString(), qbCt, rbCt, wrCt, teCt, dCt, kCt, benchCt, Flex())

    private fun setId(id: String) {
        this.id = id
    }

    val numStartingPositions: Int
        get() = validPositions.size

    val rosterSize: Int
        get() {
            var size = qbCount + rbCount + wrCount + teCount + dstCount + kCount + benchCount
            if (flex != null) {
                size += flex!!.qbrbwrteCount + flex!!.rbwrteCount + flex!!.rbteCount +
                        flex!!.rbwrCount + flex!!.wrteCount
            }
            return size
        }

    fun getNumberStartedOfPos(position: String?): Int {
        var total = 0
        when (position) {
            Constants.QB -> {
                total = qbCount
                if (flex != null) {
                    // Assume all qbs
                    total += flex!!.qbrbwrteCount
                }
            }
            Constants.RB -> {
                total = rbCount
                if (flex != null) {
                    total += flex!!.rbteCount
                    total += flex!!.rbwrCount
                    total += flex!!.rbwrteCount
                }
            }
            Constants.WR -> {
                total = wrCount
                if (flex != null) {
                    total += flex!!.wrteCount
                    total += flex!!.rbwrCount
                    total += flex!!.rbwrteCount
                }
            }
            Constants.TE -> total = teCount
            Constants.DST -> total = dstCount
            Constants.K -> total = kCount
        }
        return total
    }

    fun isPositionValid(pos: String?): Boolean {
        return validPositions.contains(pos)
    }

    class Flex @JvmOverloads constructor(rbwr: Int = Constants.ONE_STARTER, rbwrte: Int = Constants.NO_STARTERS,
                                         rbte: Int = Constants.NO_STARTERS, wrte: Int = Constants.NO_STARTERS,
                                         qbrbwrte: Int = Constants.NO_STARTERS) {
        var rbwrCount = 0
        var rbwrteCount = 0
        var rbteCount = 0
        var wrteCount = 0
        var qbrbwrteCount = 0

        init {
            rbwrCount = rbwr
            rbteCount = rbte
            rbwrteCount = rbwrte
            wrteCount = wrte
            qbrbwrteCount = qbrbwrte
        }
    }

    init {
        setId(id)
        this.qbCount = qbCt
        this.rbCount = rbCt
        this.wrCount = wrCt
        this.teCount = teCt
        this.dstCount = dCt
        this.kCount = kCt
        this.benchCount = benchCt
        this.flex = flex
        this.validPositions = HashSet()
        if (qbCt > 0 || flex.qbrbwrteCount > 0) {
            validPositions.add(Constants.QB)
        }
        if (rbCt > 0 || flex.rbwrCount > 0 || flex.rbwrteCount > 0 || flex.rbteCount > 0 || flex.qbrbwrteCount > 0) {
            validPositions.add(Constants.RB)
        }
        if (wrCt > 0 || flex.rbwrCount > 0 || flex.rbwrteCount > 0 || flex.wrteCount > 0 || flex.qbrbwrteCount > 0) {
            validPositions.add(Constants.WR)
        }
        if (teCt > 0 || flex.rbwrteCount > 0 || flex.rbteCount > 0 || flex.wrteCount > 0 || flex.qbrbwrteCount > 0) {
            validPositions.add(Constants.TE)
        }
        if (dCt > 0) {
            validPositions.add(Constants.DST)
        }
        if (kCt > 0) {
            validPositions.add(Constants.K)
        }
    }
}